File: generator/src/main/java/ch/bailu/gtk/converter/AliasTable.java
package ch.bailu.gtk.converter;

import java.util.HashMap;
import java.util.Map;

import ch.bailu.gtk.tag.AliasTag;

public class AliasTable {
    private final Map<NamespaceType, NamespaceType> table = new HashMap<>();


    private final static AliasTable INSTANCE = new AliasTable();

    public static AliasTable instance() {
        return INSTANCE;
    }

    private AliasTable() {
        NamespaceType from = new NamespaceType("glib", "String");
        NamespaceType to = new NamespaceType("glib", "GString");
        add(from, to);
    }

    public void add(NamespaceType from, NamespaceType to) {
        table.put(from, to);
    }


    public NamespaceType convert(NamespaceType from) {
        NamespaceType to = table.get(from);
        if (to == null) {
            to = from;
        }
        return to;
    }

    public void add(String namespace, String fromName, String toName) {
        var from = new NamespaceType(namespace, fromName);
        var to = new NamespaceType(namespace, toName);
        add(from, to);
    }


}

File: generator/src/main/kotlin/ch/bailu/gtk/table/AliasTable.kt
package ch.bailu.gtk.table

import ch.bailu.gtk.log.Logable
import ch.bailu.gtk.model.type.NamespaceType
import java.io.Writer

object AliasTable : Logable {
    val table: MutableMap<NamespaceType, NamespaceType> = HashMap()

    init {
        add("glib", "String", "GString")
    }


    fun add(from: NamespaceType, to: NamespaceType) {
        table[from] = to
    }

    fun add(namespace: String, fromName: String, toName: String) {
        val from = NamespaceType(namespace, fromName)
        val to = NamespaceType(namespace, toName)
        add(from, to)
    }

    fun convert(fallbackNamespace: String, typeName: String): NamespaceType {
        return convert(NamespaceType(fallbackNamespace, typeName))
    }

    fun convert(from: NamespaceType): NamespaceType {
        return table[from] ?: from
    }

    override fun log(writer: Writer) {
        writer.write("# ${AliasTable.javaClass.canonicalName}\n\n")
        writer.write("| From                           | To\n")
        writer.write("|--------------------------------|---\n")
        table.forEach {
            writer.write(String.format("| %-30s | %s\n", it.key, it.value))
        }
    }
}
